File: Backend/app/Models/Student.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Student extends Model
{
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'student_id',
        'user_id',
        'major_id',
        'name',
        'email',
        'avatar_url',
        'year_level',
        'gpa',
        'credits_taken',
        'total_credits',
        'tuition_fees',
        'academic_year',
        'class',
    ];

    protected $casts = [
        'major_id' => 'integer',
        'year_level' => 'integer',
        'gpa' => 'decimal:2',
        'credits_taken' => 'integer',
        'total_credits' => 'integer',
        'tuition_fees' => 'decimal:3',
    ];

    protected $appends = [
        'credits_progress_percentage',
        'outstanding_balance',
    ];

    /**
     * Get the user associated with the student.
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Get the major (program) the student is enrolled in.
     */
    public function major(): BelongsTo
    {
        return $this->belongsTo(Major::class);
    }

    /**
     * Get student courses (enrollments with grades).
     */
    public function studentCourses()
    {
        return $this->hasMany(StudentCourse::class);
    }

    /**
     * Get attendance records.
     */
    public function attendances()
    {
        return $this->hasMany(StudentAttendance::class);
    }

    /**
     * Get the courses the student is enrolled in.
     */
    public function courses(): BelongsToMany
    {
        return $this->belongsToMany(Course::class, 'student_courses')
            ->withPivot([
                'program_course_id',
                'year_taken',
                'semester_taken',
                'cc_score',
                'ds_score',
                'exam_score',
                'cc_weight',
                'ds_weight',
                'exam_weight',
                'final_grade',
                'letter_grade',
                'status'
            ])
            ->withTimestamps();
    }

    /**
     * Get all enrolled courses (active enrollment).
     */
    public function enrolledCourses(): BelongsToMany
    {
        return $this->courses()->wherePivot('status', 'enrolled');
    }

    /**
     * Get all attendance records.
     */
    public function attendanceRecords(): HasMany
    {
        return $this->hasMany(AttendanceRecord::class);
    }

    /**
     * Get all bills for the student.
     */
    public function bills(): HasMany
    {
        return $this->hasMany(Bill::class);
    }

    /**
     * Get all payments made by the student.
     */
    public function payments(): HasMany
    {
        return $this->hasMany(Payment::class);
    }

    /**
     * Get all event registrations.
     */
    public function eventRegistrations(): HasMany
    {
        return $this->hasMany(EventRegistration::class);
    }

    /**
     * Get events the student is registered for.
     */
    public function events(): BelongsToMany
    {
        return $this->belongsToMany(Event::class, 'event_registrations')
            ->withPivot(['registered_at', 'status'])
            ->withTimestamps();
    }

    /**
     * Get all club memberships.
     */
    public function clubMemberships(): HasMany
    {
        return $this->hasMany(ClubMembership::class);
    }

    /**
     * Get clubs the student is a member of.
     */
    public function clubs(): BelongsToMany
    {
        return $this->belongsToMany(Club::class, 'club_memberships')
            ->withPivot(['join_date', 'role', 'status'])
            ->withTimestamps();
    }

    /**
     * Get credits progress percentage.
     */
    public function getCreditsProgressPercentageAttribute(): float
    {
        if ($this->total_credits == 0) {
            return 0;
        }
        return round(($this->credits_taken / $this->total_credits) * 100, 2);
    }

    /**
     * Get outstanding balance (pending bills).
     */
    public function getOutstandingBalanceAttribute(): float
    {
        return $this->bills()->where('status', 'pending')->sum('amount');
    }

    /**
     * Calculate overall attendance percentage.
     */
    public function getAttendancePercentage(): float
    {
        $totalRecords = $this->attendanceRecords()->count();
        if ($totalRecords == 0) {
            return 0;
        }
        $presentRecords = $this->attendanceRecords()->where('status', 'present')->count();
        return round(($presentRecords / $totalRecords) * 100, 2);
    }

    /**
     * Get current semester courses.
     */
    public function getCurrentSemesterCourses()
    {
        return $this->enrolledCourses()
            ->whereHas('academicCalendar', function ($query) {
                $query->where('status', 'active');
            })
            ->get();
    }

    /**
     * Get all grades across all years (full transcript).
     */
    public function getFullTranscript()
    {
        return $this->studentCourses()
            ->with(['course', 'programCourse'])
            ->orderBy('year_taken')
            ->orderBy('semester_taken')
            ->get();
    }

    /**
     * Get transcript for a specific year.
     */
    public function getTranscriptByYear(int $year)
    {
        return $this->studentCourses()
            ->with(['course', 'programCourse'])
            ->where('year_taken', $year)
            ->orderBy('semester_taken')
            ->get();
    }

    /**
     * Get transcript for a specific year and semester.
     */
    public function getTranscriptByYearAndSemester(int $year, int $semester)
    {
        return $this->studentCourses()
            ->with(['course', 'programCourse'])
            ->where('year_taken', $year)
            ->where('semester_taken', $semester)
            ->get();
    }

    /**
     * Calculate GPA across all completed courses.
     */
    public function calculateOverallGPA(): float
    {
        $completedCourses = $this->studentCourses()
            ->where('status', 'completed')
            ->whereNotNull('final_grade')
            ->with('course')
            ->get();

        if ($completedCourses->isEmpty()) {
            return 0.0;
        }

        $totalPoints = 0;
        $totalCredits = 0;

        foreach ($completedCourses as $enrollment) {
            $gradePoint = $enrollment->getGradePoint();
            $credits = $enrollment->course->credits;
            $totalPoints += $gradePoint * $credits;
            $totalCredits += $credits;
        }

        return $totalCredits > 0 ? round($totalPoints / $totalCredits, 2) : 0.0;
    }

    /**
     * Calculate GPA for a specific year.
     */
    public function calculateYearGPA(int $year): float
    {
        $courses = $this->studentCourses()
            ->where('year_taken', $year)
            ->where('status', 'completed')
            ->whereNotNull('final_grade')
            ->with('course')
            ->get();

        if ($courses->isEmpty()) {
            return 0.0;
        }

        $totalPoints = 0;
        $totalCredits = 0;

        foreach ($courses as $enrollment) {
            $gradePoint = $enrollment->getGradePoint();
            $credits = $enrollment->course->credits;
            $totalPoints += $gradePoint * $credits;
            $totalCredits += $credits;
        }

        return $totalCredits > 0 ? round($totalPoints / $totalCredits, 2) : 0.0;
    }

    /**
     * Get courses available for the student's current year level.
     */
    public function getAvailableCourses()
    {
        return $this->major->getCoursesByYear($this->year_level)->get();
    }

    /**
     * Get required courses for student's current year that aren't completed.
     */
    public function getRemainingRequiredCourses()
    {
        $completedCourseIds = $this->studentCourses()
            ->where('status', 'completed')
            ->pluck('course_id')
            ->toArray();

        return $this->major->getRequiredCourses()
            ->whereNotIn('courses.id', $completedCourseIds)
            ->get();
    }

    /**
     * Get all teachers for student's current year courses.
     */
    public function getCurrentYearTeachers()
    {
        return Teacher::whereHas('programCourses', function ($query) {
            $query->where('major_id', $this->major_id)
                ->where('year_level', $this->year_level);
        })->with(['programCourses' => function ($query) {
            $query->where('major_id', $this->major_id)
                ->where('year_level', $this->year_level)
                ->with('course');
        }])->get();
    }

    /**
     * Get attendance for a specific course.
     */
    public function getCourseAttendance(int $courseId)
    {
        $scheduleIds = Schedule::whereHas('programCourse', function ($query) use ($courseId) {
            $query->where('course_id', $courseId);
        })->pluck('id');

        $total = $this->attendances()->whereIn('schedule_id', $scheduleIds)->count();
        $present = $this->attendances()
            ->whereIn('schedule_id', $scheduleIds)
            ->whereIn('status', ['present', 'late'])
            ->count();

        return [
            'total' => $total,
            'present' => $present,
            'percentage' => $total > 0 ? round(($present / $total) * 100, 2) : 0,
        ];
    }
}

File: frontEnd/lib/core/models/student.dart
// ============================================================================
// STUDENT MODEL - Data Structure
// ============================================================================

/// Major/Program data model
class Major {
  final int id;
  final String code;
  final String name;
  final String department;
  final int durationYears;
  final int totalCreditsRequired;
  final String? degreeType;

  Major({
    required this.id,
    required this.code,
    required this.name,
    required this.department,
    required this.durationYears,
    required this.totalCreditsRequired,
    this.degreeType,
  });

  factory Major.fromJson(Map<String, dynamic> json) {
    return Major(
      id: json['id'] as int,
      code: json['code'] as String,
      name: json['name'] as String,
      department: json['department'] as String,
      durationYears: json['duration_years'] as int,
      totalCreditsRequired: json['total_credits_required'] as int,
      degreeType: json['degree_type'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'name': name,
      'department': department,
      'duration_years': durationYears,
      'total_credits_required': totalCreditsRequired,
      'degree_type': degreeType,
    };
  }
}

/// Student data model
class Student {
  final int id;
  final String studentId;
  final String name;
  final String email;
  final int? majorId;
  final Major? major;
  final int yearLevel;
  final String? academicYear;
  final String? classLevel;
  final double? gpa;
  final int? creditsTaken;
  final int? totalCredits;
  final double? tuitionFees;
  final String? avatarUrl;

  Student({
    required this.id,
    required this.studentId,
    required this.name,
    required this.email,
    this.majorId,
    this.major,
    required this.yearLevel,
    this.academicYear,
    this.classLevel,
    this.gpa,
    this.creditsTaken,
    this.totalCredits,
    this.tuitionFees,
    this.avatarUrl,
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    print('🔍 Student.fromJson: Parsing student data...');
    print('🔍 Student.fromJson: id = ${json['id']} (${json['id'].runtimeType})');
    print('🔍 Student.fromJson: year_level = ${json['year_level']} (${json['year_level'].runtimeType})');
    print('🔍 Student.fromJson: credits_taken = ${json['credits_taken']} (${json['credits_taken']?.runtimeType})');
    print('🔍 Student.fromJson: total_credits = ${json['total_credits']} (${json['total_credits']?.runtimeType})');
    
    return Student(
      id: json['id'] as int,
      studentId: json['student_id'] as String,
      name: json['name'] as String,
      email: json['email'] as String,
      majorId: json['major_id'] as int?,
      major: json['major'] != null ? Major.fromJson(json['major'] as Map<String, dynamic>) : null,
      yearLevel: json['year_level'] as int,
      academicYear: json['academic_year'] as String?,
      classLevel: json['class'] as String?,
      gpa: json['gpa'] != null ? (json['gpa'] is String ? double.tryParse(json['gpa']) : (json['gpa'] as num).toDouble()) : null,
      creditsTaken: json['credits_taken'] != null ? (json['credits_taken'] is int ? json['credits_taken'] : int.tryParse(json['credits_taken'].toString())) : null,
      totalCredits: json['total_credits'] != null ? (json['total_credits'] is int ? json['total_credits'] : int.tryParse(json['total_credits'].toString())) : null,
      tuitionFees: json['tuition_fees'] != null ? (json['tuition_fees'] is String ? double.tryParse(json['tuition_fees']) : (json['tuition_fees'] as num).toDouble()) : null,
      avatarUrl: json['avatar_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'student_id': studentId,
      'name': name,
      'email': email,
      'major_id': majorId,
      'major': major?.toJson(),
      'year_level': yearLevel,
      'academic_year': academicYear,
      'class': classLevel,
      'gpa': gpa,
      'credits_taken': creditsTaken,
      'total_credits': totalCredits,
      'tuition_fees': tuitionFees,
      'avatar_url': avatarUrl,
    };
  }

  String get majorName => major?.name ?? 'Unknown Major';
  String get majorCode => major?.code ?? 'N/A';
  int get creditsRemaining => (totalCredits ?? 0) - (creditsTaken ?? 0);
}
